crypto/engine/eng_list.c

```diff
@@ -415,6 +437,30 @@ ENGINE *ENGINE_by_id(const char *id)
     if (strcmp(id, "dynamic")) {
         if ((load_dir = ossl_safe_getenv("OPENSSL_ENGINES")) == NULL)
             load_dir = ENGINESDIR;
+
+        load_dir = OPENSSL_strdup(load_dir);
+
+        if (stat(load_dir, &s) == -1) {
+            /* try out relative path to shared lib with our symbol */
+            if (dladdr(OpenSSL_version, &info)) {
```

Comment:
If we do wind up doing this, all of these locations should be consolidated into a single function that gets the base path for the directory you are looking for so that it can be implemented for other platforms more easily